README.md
# pilz_templates
Templates and example applications for the PRBT manipulator.

## `prbt_default_application` package

Copy the launch file into your application package. If you want to use
all configuration options, download this repository and use the files
as template.

Please see the [tutorials](https://wiki.ros.org/pilz_robots/Tutorials/) 
for explanation of the directory structure. Links inside the
launch file point to more information about individual options.

For information about the python setup please see [this thread on answer.ros.org](https://answers.ros.org/question/256611/catkin-rospy-package-structure/) or the [python style guide](http://wiki.ros.org/PyStyleGuide)

It should not be necessary, to change the original launch files
`robot.launch` or `move_group.launch` and thus fork pilz_robots.
If there are any options missing, please file an Issue.

### Versions
`pilz_templates` work both with *ROS melodic* and *ROS kinetic*.
There is a single `master` branch for both releases.

### Changelog
Changes are listed in [Changelog](prbt_default_application/CHANGELOG.rst).
If options changed in a recent version, please update your working copies
accordingly.


prbt_default_application/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package prbt_default_application
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This file lists the changes that you have to trace, once you created an own application based on this templates
and update its dependencies. Date and upstream versions are given for reference, you can check for your local
version using ``dpkg -l | grep "prbt-\|pilz-"``.

Forthcoming
-----------
* compatible with pilz_robots version 0.5.22 (melodic) or 0.6.0 (noetic) and pilz_industrial_motion version 0.4.13 (melodic) or 0.5.0 (noetic)
* changing references from pilz_command_planner to pilz_industrial_motion_planner
* Contributors: Pilz GmbH and Co. KG

2020-05-03
-----------
* compatible with pilz_robots version melodic-0.5.15
* add launchfile argument ``iso10218_support``
* update launchfile to new version in prbt_support: rename ``sto`` to ``safety_hw`` and ``sto_modbus_server_ip`` to ``modbus_server_ip``
* add options to disable operation mode and braketest features
* Contributors: Pilz GmbH and Co. KG

2020-02-17
----------
* compatible with pilz_robots version melodic-0.5.12
* include ``fake_safety_interface.launch`` in case of ``sim:=True``

2019-06-03
----------
* compatible with pilz_robots melodic version 0.5.4 / kinetic version 0.4.9
* instantiate gripper macro (see https://github.com/PilzDE/pilz_robots/pull/130)
* example how to use PG70 gripper with brackets

2019-04-10
----------
* Add template package with default launch file and configuration examples
* Contributors: Pilz GmbH and Co. KG


prbt_default_application/scripts/myFirstApplication.py
#!/usr/bin/env python
from geometry_msgs.msg import Pose, Point
from pilz_robot_programming import *
import math
import rospy

__REQUIRED_API_VERSION__ = "1"    # API version
__ROBOT_VELOCITY__ = 0.5          # velocity of the robot

# main program
def start_program():

    rospy.loginfo("Program started") # log

    # important positions
    start_pos = [1.49, -0.54, 1.09, 0.05, 0.91,-1.67]   # joint values

    pick_pose = Pose(position=Point (0, -0.5, 0.25), orientation=from_euler(0, math.radians(180), math.radians(0))) # cartesian coordinates
    work_station_pose = Pose(position=Point(-0.5, 0.1, 0.2) , orientation=from_euler(0, math.radians(-135), math.radians(90)))  # cartesian coordinates
    place_pose = Pose(position=Point(-0.1,0.4,0.25) , orientation=from_euler(0, math.radians(180),  math.radians(90))) # cartesian coordinates

    # move to start point with joint values to avoid random trajectory
    r.move(Ptp(goal=start_pos, vel_scale=__ROBOT_VELOCITY__))

    rospy.loginfo("Start loop") # log
    for _ in range(5):
        # 5 times

        # pick the PNOZ
        rospy.loginfo("Move to pick position") # log
        r.move(Ptp(goal=pick_pose, vel_scale = __ROBOT_VELOCITY__, relative=False))
        rospy.loginfo("Pick movement") # log
        pick_and_place()

        # put the PNOZ in a "machine"
        rospy.loginfo("Move to virtual machine") # log
        r.move(Ptp(goal=work_station_pose,vel_scale = __ROBOT_VELOCITY__, relative=False))
        rospy.loginfo("Place PNOZ in machine") # log
        pick_and_place()
        rospy.sleep(1)      # Sleeps for 1 sec (wait until work is finished)
        rospy.loginfo("Pick PNOZ from machine") # log
        pick_and_place()

        # place the PNOZ
        rospy.loginfo("Move to place position") # log
        r.move(Ptp(goal=place_pose, vel_scale = __ROBOT_VELOCITY__, relative=False))
        rospy.loginfo("Place movement") # log
        pick_and_place()

def pick_and_place():
    """pick and place function"""

    # a static velocity of 0.2 is used
    # the position is given relative to the TCP.
    r.move(Lin(goal=Pose(position=Point(0, 0, 0.1)), reference_frame="prbt_tcp", vel_scale=0.2))
    rospy.loginfo("Open/Close the gripper") # log
    rospy.sleep(0.2)    # pick or Place the PNOZ (close or open the gripper)
    r.move(Lin(goal=Pose(position=Point(0, 0, -0.1)), reference_frame="prbt_tcp", vel_scale=0.2))


if __name__ == "__main__":
    # init a rosnode
    rospy.init_node('robot_program_node')

    # initialisation
    r = Robot(__REQUIRED_API_VERSION__)  # instance of the robot

    # start the main program
    start_program()


prbt_default_application/scripts/myApplication.py
#!/usr/bin/env python
from geometry_msgs.msg import Pose, Point
from pilz_robot_programming import Lin, Ptp, Robot
import math
import rospy
from tf.transformations import quaternion_from_euler

__REQUIRED_API_VERSION__ = "1"    # API version
__ROBOT_VELOCITY__ = 0.5          # velocity of the robot

# important positions
start_pos = [1.49, -0.54, 1.09, 0.05, 0.91, -1.67]   # joint values


pick_pose = Pose(position=Point(0, -0.5, 0.25),  # cartesian coordinates
                 orientation=quaternion_from_euler(
                   0,
                   math.radians(180),
                   math.radians(0)))
work_station_pose = Pose(position=Point(-0.5, 0.1, 0.2),
                         orientation=quaternion_from_euler(
                           0,
                           math.radians(-135),
                           math.radians(90)))
place_pose = Pose(position=Point(-0.1, 0.4, 0.25),
                  orientation=quaternion_from_euler(
                    0,
                    math.radians(180),
                    math.radians(90)))

# main program
def start_program():
    rospy.loginfo("Program started")  # log

    # move to start point with joint values to avoid random trajectory
    r.move(Ptp(goal=start_pos, vel_scale=__ROBOT_VELOCITY__))

    rospy.loginfo("Start loop")  # log
    while not rospy.is_shutdown():
        # do infinite loop

        # pick the PNOZ
        rospy.loginfo("Move to pick position")  # log
        r.move(Ptp(goal=pick_pose,
                   vel_scale=__ROBOT_VELOCITY__,
                   relative=False))
        rospy.loginfo("Pick movement")  # log
        pick_and_place()

        # put the PNOZ in a "machine"
        rospy.loginfo("Move to virtual machine")  # log
        r.move(Ptp(goal=work_station_pose,
                   vel_scale=__ROBOT_VELOCITY__,
                   relative=False))
        rospy.loginfo("Place PNOZ in machine")  # log
        pick_and_place()
        rospy.sleep(1)      # Sleeps for 1 sec (wait until work is finished)
        rospy.loginfo("Pick PNOZ from machine")  # log
        pick_and_place()

        # place the PNOZ
        rospy.loginfo("Move to place position")  # log
        r.move(Ptp(goal=place_pose,
                   vel_scale=__ROBOT_VELOCITY__,
                   relative=False))
        rospy.loginfo("Place movement")  # log
        pick_and_place()


def pick_and_place():
    """pick and place function"""

    # a static velocity of 0.2 is used
    # the position is given relative to the TCP.
    r.move(Lin(goal=Pose(position=Point(0, 0, 0.1)),
               reference_frame="prbt_tcp",
               vel_scale=0.2))
    rospy.loginfo("Open/Close the gripper")  # log
    rospy.sleep(0.2)    # pick or Place the PNOZ (close or open the gripper)
    r.move(Lin(goal=Pose(position=Point(0, 0, -0.1)),
               reference_frame="prbt_tcp",
               vel_scale=0.2))


if __name__ == "__main__":
    # init a rosnode
    rospy.init_node('robot_program_node')

    # initialisation
    r = Robot(__REQUIRED_API_VERSION__)  # instance of the robot

    # start the main program
    start_program()
